arangod/RocksDBEngine/RocksDBSyncThread.cpp

```diff
@@ -0,0 +1,140 @@
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBSyncThread.h"
+#include "Basics/ConditionLocker.h"
+#include "Basics/RocksDBUtils.h"
+#include "Logger/Logger.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+
+#include <rocksdb/status.h>
+#include <rocksdb/utilities/transaction_db.h>
+
+using namespace arangodb;
+
+RocksDBSyncThread::RocksDBSyncThread(RocksDBEngine* engine, uint64_t interval)
+    : Thread("RocksDBSync"), 
+      _engine(engine), 
+      _interval(interval),
+      _lastSyncTime(std::chrono::steady_clock::now()),
+      _lastSequenceNumber(0) {}
+
+RocksDBSyncThread::~RocksDBSyncThread() { shutdown(); }
+ 
+Result RocksDBSyncThread::syncWal() { 
+  // note the following line in RocksDB documentation (rocksdb/db.h):
+  // > Currently only works if allow_mmap_writes = false in Options.
+  TRI_ASSERT(!_engine->rocksDBOptions().allow_mmap_writes);
+      
+  auto db = _engine->db()->GetBaseDB();
+   
+  // set time of last syncing under the lock
+  auto const now = std::chrono::steady_clock::now();
+  {
+    CONDITION_LOCKER(guard, _condition);
+
+    if (now > _lastSyncTime) {
+      // update last sync time...
+      _lastSyncTime = now;
+    }
+        
+    auto lastSequenceNumber = db->GetLatestSequenceNumber();
+  
+    if (lastSequenceNumber > _lastSequenceNumber) {
+      // update last sequence number
+      _lastSequenceNumber = lastSequenceNumber;
+    }
+  }
+
+  // actual syncing is done without holding the lock
+  return sync(db);
+}
+  
+Result RocksDBSyncThread::sync(rocksdb::DB* db) {
+  LOG_TOPIC(TRACE, Logger::ROCKSDB) << "syncing RocksDB WAL";
+
+  rocksdb::Status status = db->SyncWAL();
+  if (!status.ok()) {
+    return rocksutils::convertStatus(status);
+  } 
+  return Result();
+}
+
+void RocksDBSyncThread::beginShutdown() {
+  Thread::beginShutdown();
+
+  // wake up the thread that may be waiting in run()
+  CONDITION_LOCKER(guard, _condition);
+  guard.broadcast();
+}
+
+void RocksDBSyncThread::run() {
+  TRI_ASSERT(_engine != nullptr);
+  auto db = _engine->db()->GetBaseDB();
+        
+  LOG_TOPIC(TRACE, Logger::ROCKSDB) << "starting RocksDB sync thread with interval " << _interval << " microseconds";
+
+  while (!isStopping()) {
+    try {
+      auto const now = std::chrono::steady_clock::now();
+     
+      {
+        // wait for time to elapse, and after that update last sync time 
+        CONDITION_LOCKER(guard, _condition);
+
+        auto const previousLastSequenceNumber = _lastSequenceNumber;
+        auto const previousLastSyncTime = _lastSyncTime;
+        auto const end = _lastSyncTime + std::chrono::microseconds(_interval);
+        if (end > now) {
+          guard.wait(std::chrono::microseconds(std::chrono::duration_cast<std::chrono::microseconds>(end - now)));
```

Comment:
should you check `isStopping()` after waiting here ?

Reply:
I don't think this is required.
If we `continue`, then we will directly afterwards check `isStopping()` anyway.
If we don't `continue`, we will do one final sync and then check `isStopping()` afterwards.
I don't think the extra sync on shutdown will do any harm (it's probably even sensible to do it...).